core-api/src/main/java/com/sequenceiq/cloudbreak/api/endpoint/v4/clustertemplate/ClusterTemplateV4Endpoint.java

```diff
@@ -46,6 +46,13 @@
             nickname = "listClusterTemplatesByWorkspace")
     ClusterTemplateViewV4Responses list(@PathParam("workspaceId") Long workspaceId);
 
+    @GET
+    @Path("get_by_env")
+    @Produces(MediaType.APPLICATION_JSON)
+    @ApiOperation(value = ClusterTemplateOpDescription.LIST_BY_ENV, produces = MediaType.APPLICATION_JSON,
+            notes = CLUSTER_TEMPLATE_NOTES, nickname = "listClusterTemplatesByEnv")
+    ClusterTemplateViewV4Responses listByEnv(@PathParam("workspaceId") Long workspaceId, @QueryParam("environmentCrn") String environmentCrn);
```

Comment:
why do we need to introduce another endpoint for this? the env crn is a query param which can be empty (seemingly). If this param would missing, tha call would go to the simple list endpoint (line 47).
IMO we shouldn't introduce this endpoint, we just need to extend the original list endpoint with this query param. since tha param itself is not a mandatory field, this endpoint extension shouldn't break the compatibility

Reply:
I asked Gergo to do that, since original endpoint is for listing in account and this is for listing by env and these are conpletely different from authorization point of view, also this endpoint will called by env service only so it is an internal call right now (later it is possible to authorize it for an env)

Reply:
I see, but the original endpoint could work in both ways whether the env crn is given or not. If we'd like to follow your mentioned thought about exposing this new endpoint in a "public" way somewhere in the future, this exact same question could emerge again (namely: why do we have two endpoints for a seemingly quite identical behavior). And at that point, we may have to deprecate this endpoint which comes with the concern of breaking something that uses this endpoint (but we don't know about it, because it may be some experience or other embedding service).

Back to the AuthZ. What - component - decides that this given endpoint is an internal or an "external" endpoint?
I mean if the internal kind of behavior means only that we have to implement its mechanism in a different way one (or more) layer lower, I think my suggestion could have more weight.
If the decision of which service can call this - new - endpoint and which can't is in another component's hand then I have nothing to say here, it is fine to me.
This section is partly because most likely your general knowledge around our AuthN/AuthZ is much deeper than mine.

I would like to include more people into this if you don't mind to have/obtain a more diverse opinion.

Reply:
yes you are right, but we need this because of the authorization. Please talk with @horadla23

Reply:
@doktoric , @biharitomi , @daszabo could you please take a look at the abovementioned topic(s) and leave some opinion about it? thanks guys!

Reply:
ok so my point here is that even both of them are identical from developer/code point of view but it is different from user point of view. Original list gives all templates in the account/tenant, this new one gives them back by an environment . This is different from customer point of view and since we are authorizing calls on api, it is completely different from authz point of view. one of them should be authorized based on account and this new one should be authorized on env level. we had so much pain caused by these mixed endpoints during implementation of authorization so we should avoid mixing these slightly different calls into 1 method.

you can check readme of authorization-common module about this topic and also we will hold a demo about it.

we can involve team into this discussion of course.

(so this is not about that we need to differentiate them because one of them is internal, rather it is about api design which considers different kind of authorizations)

Reply:
what if the env crn would be as a path param? Would it be feasible for both side? 😄  @horadla23 @gregito cc: @topolyai5 
Of course this endpoint's responses would be the subset of the list by tenant, but I think it is more acceptable this way.

Reply:
using the crn as a path param would emphasize the difference between two endpoints thus this approach would be more acceptable to me 👍🏻

Reply:
I am also fine with an optional path param. One more thing If we add this ned function probably @andrasfelfoldi and @zoltanzseli can use this in case of DH creation page. This is a kind of optimization.

Reply:
I don't support the env crn in the path, because we want to filter instead of separate the resource by env. I think the best solution an optional param on the list operation but this is a problem for authz team.

Reply:
if UI will use this, then that is one more reason to think about how UI will use it and how we will authorize this.